map_model/src/make/buildings.rs

```diff
@@ -360,21 +363,81 @@ fn classify_bldg(
     tags: &BTreeMap<String, String>,
     amenities: &BTreeSet<(String, String)>,
     area_sq_meters: f64,
+    rng: &mut rand_xorshift::XorShiftRng,
 ) -> BuildingType {
+    // used: top values from https://taginfo.openstreetmap.org/keys/building#values (>100k uses)
     let tags = Tags(tags);
 
+    let mut commercial = false;
+    let mut workers = 0;
+ 
     // These are (name, amenity type) pairs, produced by get_bldg_amenities in
     // convert_osm/src/osm_reader.rs.
     if !amenities.is_empty() {
-        return BuildingType::Commercial;
+        commercial = true;
     }
 
-    if tags.is("key", "value") {
-        return BuildingType::Residential(42);
+    if tags.is("ruins", "yes") {
+        if commercial {
+            return BuildingType::Commercial;
+        }
+        return BuildingType::Empty;
     }
 
-    // 1 person per 10 square meters
-    BuildingType::Residential((area_sq_meters / 10.0) as usize)
+    if tags.is_any("building", vec!["office", "industrial", "commercial", "retail", "warehouse", "civic", "public"]) {
+        return BuildingType::Commercial;
+    } else if tags.is_any("building", vec!["school", "university", "construction", "church"]) {
+        // TODO: special handling in future
+        return BuildingType::Empty;
+    } else if tags.is_any("building", vec!["garage", "garages", "shed", "roof", "greenhouse", "farm_auxiliary", "barn", "service"]) {
+        return BuildingType::Empty;
+    } else if tags.is_any("building", vec!["house", "detached", "semidetached_house", "farm"]) {
+        workers = rng.gen_range(0, 3);
+    } else if tags.is_any("building", vec!["hut", "static_caravan", "cabin"]) {
+        workers = rng.gen_range(0, 2);
+    } else if tags.is_any("building", vec!["apartments", "terrace", "residential"]) {
+        let mut levels = 1;
+        // TODO: replace by a proper getter
```

Comment:
`let levels = tags.0.get("building:levels").map(|x| x.parse::<usize>().ok()).unwrap_or(1)`

Bit of a mouthful. Breaking down slightly:
- `tags.0` just grabs the `BTreeMap<String, String>`; `Tags` is a "wrapper" struct to add a new convenience method
- `get("key")` returns `Option<&String>`, since it might not exist
- ` map` transforms the case that the Option is Some, leaves the None case alone
- `x.parse::<usize>()` returns a `Result<usize, some error type>`, but we don't care about the parsing error, so `.ok()` makes it an `Option<usize>`

Reply:
`expected enum `std::option::Option`, found integer`

After shotgun debugging I ended with `let levels = tags.0.get("building:levels").map(|x| x.parse::<usize>().ok()).unwrap_or(Some(1)).unwrap();` that compiles and maybe will work.

Reply:
Woops, should be `let levels = tags.0.get("building:levels").and_then(|x| x.parse::<usize>().ok()).unwrap_or(1)`

x.parse().ok() returns Option<usize>, and we don't want Option<Option<usize>>, so `and_then` does the flattening. Hopefully you're not sinking too much time into the Rust bits

Reply:
No, I try quick googling and start asking for help here. I had vague plans to learn Rust a bit so I am not considering it as a wasted time anyway.

BTW, why in `sim/src/make/generator.rs`

```
                let mut deterministicRng = XorShiftRng::seed_from_u64(home.0 as u64);
                let work = *workplaces.choose(deterministicRng).unwrap();
```

gets hit with `function or associated item not found in \`rand_xorshift::XorShiftRng\``?

I coped that code from `map_model/src/make/buildings.rs` where it is working, .toml seems to have the same version...